src/python/pants/backend/python/rules/resolve_requirements.py

```diff
@@ -0,0 +1,90 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from future.utils import text_type
+
+from pants.backend.python.subsystems.python_native_code import PexBuildEnvironment, PythonNativeCode
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.engine.fs import Digest, Snapshot, UrlToFetch
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.rules import optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.util.objects import datatype, hashable_string_list, string_optional, string_type
+from pants.util.strutil import create_path_env_var
+
+
+class ResolveRequirementsRequest(datatype([
+  ('requirements', hashable_string_list),
+  ('output_filename', string_type),
+  ('entry_point', string_optional),
+  ('interpreter_constraints', hashable_string_list),
+])):
+  pass
+
+
+class ResolvedRequirementsPex(datatype([
+  ('directory_digest', Digest),
+  ('requirements', hashable_string_list)
+])):
+  pass
+
+
+# TODO: This is non-hermetic because the requirements will be resolved on the fly by
+# pex, where it should be hermetically provided in some way.
+@rule(ResolvedRequirementsPex, [ResolveRequirementsRequest, PythonSetup, PexBuildEnvironment])
+def resolve_requirements(request, python_setup, pex_build_environment):
+  """Returns a PEX with the given requirements, optional entry point, and optional
+  interpreter constraints."""
+
+  # Sort all user requirement strings to increase the chance of cache hits across invocations.
```

Comment:
As @illicitonion pointed out above, this actually needs to happen outside of the rule in order for the `@rule` memoization to kick in. That's a slightly unfortunate thing from an API perspective... makes me wonder whether a `sorted` constraint for datatype fields would be reasonable...

Reply:
We could add a runtime assertion in the meantime to ensure callers pre-sort. Not sure if that's worth the extra code and runtime hit?

Also, sounds like `request.interpreter_constraints` needs to be sorted already?